framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js

```diff
@@ -42,47 +45,210 @@ class BlockSynchronizationMechanism {
-			// TODO: handle bestPeer and move on to step 2 defined in
-			// https://github.com/LiskHQ/lips/blob/master/proposals/lip-0014.md#block-synchronization-mechanism
-			// ...
-
-			// The node requests the last common block C from P (Peer).
-			const { data: networkLastBlock } = await this.channel.invoke(
-				'network:requestFromPeer',
-				{
-					procedure: 'getLastBlock',
-					peerId: bestPeer.id,
-				},
-			);
 
-			const { valid: validBlock, err } = await this._blockDetachedStatus(
-				networkLastBlock,
+			const bestPeer = await this._computeBestPeer(peers);
+			await this._requestAndValidateLastBlock(receivedBlock, bestPeer);
+			await this._revertToLastCommonBlock(receivedBlock, bestPeer);
+		} finally {
+			this.active = false;
+		}
+	}
+
+	/**
+	 * Helper function that encapsulates:
+	 * 1. applying a penalty to a peer.
+	 * 2. restarting sync.
+	 * 3. throwing the reason.
+	 * @param {object} peer - The object that contains the peer ID to target
+	 * @param receivedBlock
+	 * @param {Error } error - An error object containing the reason for applying
+	 * a penalty and restarting sync
+	 * @private
+	 */
+	_applyPenaltyAndRestartSync(peer, receivedBlock, error) {
+		this.channel.invoke('network:applyPenalty', {
+			peerId: peer.id,
+			penalty: 100,
+		});
+		this.channel.publish('chain:processor:sync', { block: receivedBlock });
+		throw error;
+	}
+
+	/**
+	 * Reverts the current chain so the new tip of the chain corresponds to the
+	 * last common block.
+	 * @param receivedBlock
+	 * @param {Object} peer - The selected peer to target.
+	 * @return {Promise<void>}
+	 * @private
+	 */
+	async _revertToLastCommonBlock(receivedBlock, peer) {
+		const lastCommonBlock = await this._requestLastCommonBlock(peer);
+
+		if (!lastCommonBlock) {
+			this._applyPenaltyAndRestartSync(
+				peer,
+				receivedBlock,
+				new Error(
+					`No common block has been found between the system chain and the targeted peer ${
+						peer.id
+					}`,
+				),
 			);
-			const notInDifferentChain = ForkChoiceRule.isDifferentChain(
-				this.blocks.lastBlock,
-				networkLastBlock,
+		}
+
+		if (lastCommonBlock.height < this.bft.finalizedHeight) {
+			this._applyPenaltyAndRestartSync(
+				peer,
+				new Error(
+					'The last common block height is less than the finalized height of the current chain',
+				),
 			);
+		}
+
+		await deleteBlocksAfterHeightAndBackup(
+			this.processorModule,
+			this.blocks,
+			lastCommonBlock.height,
+		);
+	}
 
-			if (!validBlock || !notInDifferentChain) {
-				this.channel.invoke('network:applyPenalty', {
-					peerId: bestPeer.id,
-					penalty: 100,
-				});
-				this.channel.publish('chain:processor:sync', { block: receivedBlock });
-				throw err;
+	/**
+	 * Returns a list of block heights corresponding to the first block of a defined number
+	 * of rounds (listSizeLimit)
+	 *
+	 * @param listSizeLimit - The size of the array to be computed
+	 * @param _currentRound
+	 * @return {Promise<*>}
```

Comment:
does this resolve an array of blockIDs all the time? If yes, pls change to `@return {Promise<[Number]>}`

Reply:
blockIDs are string 😉